src/Features/Core/Portable/Diagnostics/BuildOnlyDiagnosticsService.cs

```diff
@@ -6,107 +6,152 @@
 using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Collections;
 using Microsoft.CodeAnalysis.Host;
 using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Diagnostics;
 
 [ExportWorkspaceServiceFactory(typeof(IBuildOnlyDiagnosticsService), ServiceLayer.Default), Shared]
 [method: ImportingConstructor]
 [method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
-internal sealed class BuildOnlyDiagnosticsServiceFactory() : IWorkspaceServiceFactory
+internal sealed class BuildOnlyDiagnosticsServiceFactory(
+    IAsynchronousOperationListenerProvider asynchronousOperationProvider) : IWorkspaceServiceFactory
 {
     public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
-        => new BuildOnlyDiagnosticsService(workspaceServices.Workspace);
+        => new BuildOnlyDiagnosticsService(workspaceServices.Workspace, asynchronousOperationProvider.GetListener(FeatureAttribute.Workspace));
 
-    private sealed class BuildOnlyDiagnosticsService : IBuildOnlyDiagnosticsService
+    private sealed class BuildOnlyDiagnosticsService : IBuildOnlyDiagnosticsService, IDisposable
     {
-        private readonly object _gate = new();
+        private readonly CancellationTokenSource _disposalTokenSource = new();
+        private readonly AsyncBatchingWorkQueue<WorkspaceChangeEventArgs> _workQueue;
+
+        private readonly SemaphoreSlim _gate = new(initialCount: 1);
         private readonly Dictionary<DocumentId, ImmutableArray<DiagnosticData>> _documentDiagnostics = [];
 
-        public BuildOnlyDiagnosticsService(Workspace workspace)
+        public BuildOnlyDiagnosticsService(
+            Workspace workspace,
+            IAsynchronousOperationListener asyncListener)
         {
+            _workQueue = new AsyncBatchingWorkQueue<WorkspaceChangeEventArgs>(
+                TimeSpan.Zero,
+                ProcessWorkQueueAsync,
+                asyncListener,
+                _disposalTokenSource.Token);
             workspace.WorkspaceChanged += OnWorkspaceChanged;
         }
 
+        public void Dispose()
+            => _disposalTokenSource.Dispose();
+
         private void OnWorkspaceChanged(object? sender, WorkspaceChangeEventArgs e)
         {
+            // Keep this switch in sync with the switch in ProcessWorkQueueAsync
```

Comment:
📝 If events that don't fit the exclusion pattern here are rare, we have the option of adding all events to the queue since it processes them quickly.